Common shared binary implementation

kishore-sn opened the pull request:
This PR contains implementation of common shared binary for TizenRT.
The common binary contains libuc, libumm, libuarch and libproxies libraries, which will be used by all the loadable apps. 

Following is working and verfied:
- Create common binary
- Load common binary and export symbols
- Import and bind symbols during app load
- Boot up and tash is working
- Mpu test, kernel tc and filesystem tc are passing


Following are pending or not verified:
- There are some global variables in shared libraries. Not verified if they can cause any issues
- Etc.. 

In order to test this, it is necessary to create a filesystem image containing the tinyara_common_binary file
and download it to the device. If this step is not done, the app loading will fail during boot up.


Size comparison of binry files (Bytes)

Case 1:
Without shared library (micom + wifi) = 				(11421 + 871857 = **883278**)
With shared library (micom + wifi + common) = 	(2881 + 744881 + 115064 = **862826**)

Case 2:
Without shared library (micom + wifi) = 			(11421 + 14185 = **25606**)
With shared library (micom + wifi + common) = (2881 + 4977 + 10916 = **18774**)

seinfra commented:
Target : [f59244c4fa78a67b9c707a13cdf88a88ad3303ba] - Code Rule Check (C++) OK.

seinfra commented:
Target : [f59244c4fa78a67b9c707a13cdf88a88ad3303ba] - Code Rule Check OK.

jeongchanKim commented on os/include/tinyara/sched.h:
```diff
@@ -631,6 +631,9 @@ struct tcb_s {
 	uint32_t uspace;		/* User space object for app binary */
 	uint32_t uheap;			/* User heap object pointer */
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+	uint8_t app_id;			/* Application id to be used to index into the app heap table in common binary */
```
We already have `binary_idx` in task_group binary_s. How about using this?

kishore-sn replied:
@jeongchanKim 
I had made the changes to use the binary_idx from tg_bininfo object. However, the tg_bininfo object is not propagated to all the tasks. When I tried to copy the tg_bininfo to the newly created tasks, then there is some issue and kernel_tc is getting stuck at various points. I am not sure why this issue is happening. 

However, i think it is better to have the app id variable directly in the tcb_s structure so that we can access it easily during a context switch. If we use tg_bininfo or tg_binid for this purpose, then getting the binary_idx would require more indirections, or calling an api to get binary_idx from tg_binid. Both these will be more time consuming and better to avoid during context switch.. So, i have reverted my previous changes and maintained the app_id variable.

thapav replied:
"i think it is better to have the app id variable directly in the tcb_s structure so that we can access it easily during a context switch"
I would agree to keep app_id in tcb_s structure. it would be more straight forward.

jeongchanKim commented on os/binfmt/libelf/libelf_bind.c:
```diff
@@ -277,6 +281,69 @@ static int elf_relocateadd(FAR struct elf_loadinfo_s *loadinfo, int relidx, FAR
+				goto ret_err;
+			}
+
+			strcpy(g_lib_symtab[g_num_lib_syms].sym_name, loadinfo->iobuffer);
```
Please use `strncpy` instead of `strcpy`.

kishore-sn replied:
We dont know the length of the iobuffer (symbol name). Also, all symbol names in the elf string table are always null terminated. So, it must be fine to perform strcpy. Also, the sym_name variable is allocated based on the `strlen(loadinfo->iobuffer) + 1`.

intom replied:
I agree with kishore-sn

jeongchanKim commented on os/binfmt/binfmt_loadlibrary.c:
```diff
@@ -0,0 +1,148 @@
+	 */
+	if (bin) {
+		if (!bin->data_backup) {
+			errcode = -EINVAL;
```
Sometimes, you set the errno with negative value, but sometimes with positive.
Please check this.

kishore-sn replied:
fixed.

seinfra commented:
Target : [9078ad090e82d6b13329b5e3af84d1fcd1ae177d] - Code Rule Check (C++) OK.

seinfra commented:
Target : [9078ad090e82d6b13329b5e3af84d1fcd1ae177d] - Code Rule Check OK.

seinfra commented:
Target : [4f41b1ffe6ebac0ffb4e4cd188e94ff666adddda] - Code Rule Check OK.

seinfra commented:
Target : [4f41b1ffe6ebac0ffb4e4cd188e94ff666adddda] - Code Rule Check (C++) OK.

kishore-sn commented:
Fixed memfault due to tg_bininfo not being copied to child task.

thapav commented on os/arch/arm/src/armv7-m/up_mpu.c:
```diff
@@ -355,3 +355,48 @@ void mpu_configure_app_regs(uint32_t *regs, uint32_t region, uintptr_t base, siz
+ *
+ * Description:
+ *   Configure the mpu region for common library
+ *
```
Can we have parameter description for the arguments ?

kishore-sn replied:
Done

thapav commented on os/binfmt/binfmt_loadlibrary.c:
```diff
@@ -119,6 +124,12 @@ int load_library(FAR const char *filename)
+#ifdef CONFIG_ARMV7M_MPU
+	mpu_configure_library_region(g_mpu_region_nr++, bin->alloc[0], bin->textsize, true, true);
+	mpu_configure_library_region(g_mpu_region_nr++, bin->alloc[3], bin->rosize, true, false);
+	mpu_configure_library_region(g_mpu_region_nr++, bin->alloc[4], bin->ramsize, false, false);
```
do you think using #defines for text ro & data indexes would be better ?

kishore-sn replied:
Done

thapav commented on os/binfmt/binfmt_loadlibrary.c:
```diff
@@ -0,0 +1,134 @@
+	return ERROR;
+
+}
+#endif							/* if defined(CONFIG_BINFMT_ENABLE) && defined(CONFIG_SUPPORT_COMMON_BINARY) */
```
This code is similar to the load_binary API in "binfmt/binfmt_loadbinary.c"
Is it necessary to keep it separate or can we merge the two with a binary/library parameter to a common API such as:
load_module(int binary_idx, FAR const char *filename, load_attr_t *load_attr)
where load_attr could be > binary/library

kishore-sn replied:
You are right. Initially i expected many changes in library loading. So, I had made new API. 
But, i feel it can be merged with loadbinary api. 

Done.

seinfra commented:
Target : [98f0b2bf95d50bd0a3aba9a8d8266e7f77963f31] - Code Rule Check (C++) OK.

seinfra commented:
Target : [98f0b2bf95d50bd0a3aba9a8d8266e7f77963f31] - Code Rule Check OK.

seinfra commented:
Target : [abde36167add5ee6dc5f127a284e63da87faac70] - Code Rule Check (C++) OK.

seinfra commented:
Target : [abde36167add5ee6dc5f127a284e63da87faac70] - Code Rule Check OK.

amogh-samsung commented on os/binfmt/binfmt_loadbinary.c:
```diff
@@ -105,28 +126,34 @@ int load_binary(int binary_idx, FAR const char *filename, load_attr_t *load_attr
 		if (!bin) {
 			berr("ERROR: Failed to allocate binary_s\n");
-			errcode = ENOMEM;
+			errcode = -ENOMEM;
```
There seems to be some inconsistency with the assignment of the error value to `errcode`
It is assigned a positive value at some places and then passed on to `set_errno` to be taken care by.
At other points, it is assigned a negative Error Signal value, eg. line 113, 168.

amogh-samsung commented on os/binfmt/libelf/libelf_bind.c:
```diff
@@ -277,6 +281,69 @@ static int elf_relocateadd(FAR struct elf_loadinfo_s *loadinfo, int relidx, FAR
+	for (i = 0; i < nsyms; i++) {
+		Elf32_Sym sym;
+		ret = elf_readsym(loadinfo, i, &sym);
+                if (ret < 0) {
```
Indentation seems to be a little off.

amandeepc commented on os/binfmt/libelf/libelf_bind.c:
```diff
@@ -277,6 +281,69 @@ static int elf_relocateadd(FAR struct elf_loadinfo_s *loadinfo, int relidx, FAR
+			goto ret_err;
+		}
+
+		if (/*ELF32_ST_TYPE(sym.st_info) == STT_FUNC &&*/ ELF32_ST_BIND(sym.st_info) == STB_GLOBAL) {
```
Comment /**/ can be removed from here

kishore-sn commented:
Added commits for below items:
- Implemented verification of udefined symbols in wifi and micom and common binary
- Implementation of new reload mechanism for crash in common binary
- Optimized loading time for common binary

tizen-build commented:
Target : [1b83fc2faf0c7e36b41c2226ec8d27d64c1cbb82] - Code Rule Check OK.

tizen-build commented:
Target : [1b83fc2faf0c7e36b41c2226ec8d27d64c1cbb82] - Code Rule Check OK.

tizen-build commented:
Target : [1b83fc2faf0c7e36b41c2226ec8d27d64c1cbb82] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [cbf61755bb2057994a653d808f80f8c5790674fa] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [cbf61755bb2057994a653d808f80f8c5790674fa] - Code Rule Check OK.

tizen-build commented:
Target : [868abad572b6733b723763ccf7774ca35896c308] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [868abad572b6733b723763ccf7774ca35896c308] - Code Rule Check OK.

jeongchanKim commented on os/Kconfig:
```diff
@@ -270,6 +270,31 @@ config APPS_RAM_REGION_SHAREABLE
+config SUPPORT_COMMON_BINARY
+	bool "Support Common Binary"
+	default n
+	depends on FS_SMARTFS
```
I think this is needed only with binary separation. How do you think about?

jeongchanKim commented on os/binfmt/libelf/libelf_bind.c:
```diff
@@ -277,6 +281,69 @@ static int elf_relocateadd(FAR struct elf_loadinfo_s *loadinfo, int relidx, FAR
+	}
+
+	binfo("Total symbols in library = %d. Exported symbols = %d\n", nsyms, g_num_lib_syms);
+	kmm_realloc(g_lib_symtab, g_num_lib_syms * sizeof(struct symtab_s));
```
Is it okay not to check that kmm_realloc fails or not?

jeongchanKim commented on os/binfmt/libelf/libelf_symbols.c:
```diff
@@ -79,6 +79,41 @@
+
+	ret = elf_read(loadinfo, loadinfo->strtab, strtab->sh_size, strtab->sh_offset);
+	if (ret != OK) {
+		berr("ERROR: Failed to load string table into memory\n");
```
Is it okay not to free the `loadinfo->strtab`?

jeongchanKim commented on os/arch/arm/src/armv7-m/up_assert.c:
```diff
@@ -462,6 +472,17 @@ static void recovery_user_assert(void)
+			uint32_t end = start + g_lib_binp->textsize;
+			if (assert_pc >= start && assert_pc <= end) {
+				/* We will set binid to zero to indicate a crash in common library */
+				binid = 0;
```
I think it can be okay for using 0, but you'd better use defined value for readability.

kishore-sn replied:
Ok. Fixed.

jeongchanKim commented on os/kernel/binary_manager/binary_manager_load.c:
```diff
@@ -271,6 +271,15 @@ static int binary_manager_load_all(void)
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+	char libname[CONFIG_NAME_MAX];
+	snprintf(libname, CONFIG_NAME_MAX, "%s%s", CONFIG_COMMON_BINARY_PATH, CONFIG_COMMON_BINARY_NAME);
+	ret = load_binary(0, libname, NULL);
```
Is it okay not to check the validation of common binary?

jeongchanKim commented on os/binfmt/libelf/libelf_bind.c:
```diff
@@ -277,6 +281,69 @@ static int elf_relocateadd(FAR struct elf_loadinfo_s *loadinfo, int relidx, FAR
+	return OK;
+
+ret_err:
+	kmm_free(g_lib_symtab);
```
Is it okay not to free the `g_lib_symtab[g_num_lib_syms].sym_name`?

kishore-sn replied:
It needs to be freed. Fixed.

tizen-build commented:
Target : [9e220829d5aa3d8d2e6fcebc73efa94b6f2658fe] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [9e220829d5aa3d8d2e6fcebc73efa94b6f2658fe] - Code Rule Check OK.

jeongchanKim commented on os/Kconfig:
```diff
@@ -270,6 +270,31 @@ config APPS_RAM_REGION_SHAREABLE
+
+config COMMON_BINARY_PATH
+	string "Path where common binary is stored"
+	default ""
```
We'd better set the initial path as `"/mnt/"`.

jeongchanKim commented:
We will support the common binary based on FS. Then I think we need to make FS image with common binary in build time. Could you add that functionality in makefile?

tizen-build commented:
Target : [389c8a175405043d60391ea825031fca03f903ec] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [389c8a175405043d60391ea825031fca03f903ec] - Code Rule Check OK.

kishore-sn commented:
Added load time optimization with hashmap implementation.

tizen-build commented:
Target : [756df4a3115e7bc1dca2f8b62cf538a5a95526a7] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [756df4a3115e7bc1dca2f8b62cf538a5a95526a7] - Code Rule Check OK.

tizen-build commented:
Target : [7a05e09d31dd2ffb095231776b337fb4f78d8734] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [7a05e09d31dd2ffb095231776b337fb4f78d8734] - Code Rule Check OK.

jeongchanKim commented on os/binfmt/libelf/libelf_symbols.c:
```diff
@@ -326,6 +322,20 @@ int elf_symvalue(FAR struct elf_loadinfo_s *loadinfo, FAR Elf32_Sym *sym, FAR co
+			return -ENOENT;
+		}
+
+		sym->st_value = (uint32_t) hashmap_get((struct hashmap_s *) exports, hashmap_get_hashval(loadinfo->iobuffer));
```
Please remove space after type casting.

tizen-build commented:
Target : [87668314c9d0d57346ff86a7434aee6f87970298] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [87668314c9d0d57346ff86a7434aee6f87970298] - Code Rule Check OK.

kishore-sn commented:
Added commit to auto generate and download smartfs image with common binary.

tizen-build commented:
Target : [bad02ad91e81c1d096ff4cde7aa9a87a7e5e0f29] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [bad02ad91e81c1d096ff4cde7aa9a87a7e5e0f29] - Code Rule Check OK.

tizen-build commented:
Target : [d1ce5217f9960edd3169ab919ecb9a3613e7e3ac] - Code Rule Check OK.

tizen-build commented:
Target : [d1ce5217f9960edd3169ab919ecb9a3613e7e3ac] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [40b2e0bea0df68716c3137e3355a8394df75a06d] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [40b2e0bea0df68716c3137e3355a8394df75a06d] - Code Rule Check OK.

tizen-build commented:
Target : [edcaaa3eb62361f1e44b2b7e97920be84f43f0fa] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [edcaaa3eb62361f1e44b2b7e97920be84f43f0fa] - Code Rule Check OK.

tizen-build commented:
Target : [0553de477895e88104529125688f74d954a4ef79] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [0553de477895e88104529125688f74d954a4ef79] - Code Rule Check OK.

tizen-build commented:
Target : [57c837b60aa2c5f43a046a6ac035a28d53e3a054] - Code Rule Check (C++) OK.

tizen-build commented:
Target : [57c837b60aa2c5f43a046a6ac035a28d53e3a054] - Code Rule Check OK.

tizen-build commented:
Target : [dca946d683b498e0dbc2406c6d564505efe47ef6] - Code Rule Check OK.

tizen-build commented:
Target : [dca946d683b498e0dbc2406c6d564505efe47ef6] - Code Rule Check (C++) OK.
